[Feat]  /login  경로 페이지 개발

yonghyeun opened the pull request:
# 관련 이슈 번호
#52
# 설명

`/login` 경로에 해당하는 페이지를 생성했습니다.

`/login` 경로에선 `features/auth/ui` 에 존재하는 `OAuthHyperlinks , EmailHyperlink` 를 이용해서 구성되었습니다. 

---

### `API` 경로 흐름 

해당 페이지에선 다음과 같은 `API` 요청 흐름이 존재합니다. 

```tsx
// features/auth/api/login.ts

interface OAuthLoginResponse {
  code: number;
  message: string;
  content: {
    token: string;
    role: string;
  };
}

export type OAuthServerName = "GOOGLE" | "NAVER";

/**
 * 해당 훅은 OAuthServerName이 주어지면 해당 OAuthServer로 로그인을 시도합니다.
 * 만약 OAuthServerName이 null일 경우엔 로그인을 시도하지 않습니다.
 */
export const useOauthLogin = (OAuthServerName: OAuthServerName | null) => {
  const isOauthLoginEnabled = OAuthServerName !== null;

  // OauthLogin을 사용 할 수 없는 경우엔 EndPoint는 false
  const END_POINT = isOauthLoginEnabled && LOGIN_END_POINT[OAuthServerName];

  // TODO 리액트 쿼리를 활용하여 최적화 하기
  const query = useQuery<OAuthLoginResponse>({
    queryKey: ["OAuthLogin", OAuthServerName],
    // TODO fetch 함수 커스텀 라이브러리로 분리하기
    queryFn: async () => {
      if (!END_POINT) {
        return;
      }

      const response = await fetch(END_POINT);
      if (!response.ok) {
        // TODO 에러 메시지 픽스하기
        throw new Error("예기치 못한 네트워크 오류가 발생했습니다.");
      }
      const data = await response.json();
      return data;
    },
    enabled: isOauthLoginEnabled,
  });

  return query;
};
```

`useOAuthLogin` 커스텀훅은 인수로 들어오는 `OauthServerName` 값에 따라 소셜 로그인을 시도합니다. 

만약 `OauthServerName` 값이 `null` 일 경우엔 패칭을 시도하지 않습니다. 


![image](https://github.com/user-attachments/assets/15b9bb12-a654-412e-94d1-0b4ee005f1e6)

```tsx
export const OAuthLoginHyperLinks = () => {
  const [OAuthServerName, setOAuthServerName] =
    useState<OAuthServerName | null>(null);

  const setToken = useAuthStore((state) => state.setToken);
  const setRole = useAuthStore((state) => state.setRole);

  const { data: authResponse } = useOauthLogin(OAuthServerName);

  useEffect(() => {
    if (authResponse) {
      const { token, role } = authResponse.content;

      setToken(token);
      setRole(role);
    }
  }, [authResponse, setToken, setRole]);

  return (
    <>
      <NaverLoginHyperLink onClick={() => setOAuthServerName("NAVER")} />
      <GoogleLoginHyperLink onClick={() => setOAuthServerName("GOOGLE")} />
    </>
  );
};
```

`OauthHyperLinks` 컴포넌트에선 위에서 정의한 커스텀훅인 `useOauthLogin` 을 호출합니다. 

첫 렌더링 시에는 `OauthServerName` 상태 값이 `null` 이기 때문에 소셜 로그인을 시도하지 않습니다.

하지만 사용자가 구글로 로그인 , 네이버로 로그인 버튼을 클릭하게 되면 `OauthServerName` 값이 설정되어 소셜 로그인을 시도하게 됩니다. 

> 다음과 같이 구현한 이유 
>
> 소셜 로그인을 시도하는 행위는 네이버 , 구글 상관없이 하나의 서버 액션이라 생각했습니다. 

----

### 스토리북 설정 변경

```tsx
import React from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { MemoryRouter } from "react-router-dom";

const queryClient = new QueryClient();

import "../src/global.css";
import type { Preview } from "@storybook/react";

const withProviders = (Story, context) => {
  const { initialEntries, initialIndex } =
    context.parameters.memoryRouter || {};

  return (
    <QueryClientProvider client={queryClient}>
      <MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
        <Story />
      </MemoryRouter>
    </QueryClientProvider>
  );
};

const preview: Preview = {
  parameters: {
    controls: {
      matchers: {
        color: /(background|color)$/i,
        date: /Date$/i,
      },
    },
  },
  // 모든 스토리에 전역적으로 react-router-dom을 적용
  decorators: [withProviders],
};

export default preview;
```

스토리북에서 리액트 라우터 돔,  리액트 쿼리의 기능을 이용하는 컴포넌트를 렌더링 해주기 위해 `preview.ts`  파일에서 `decorator` 를 설정해주었습니다.

이를 통해 모든 스토리북에서 리액트 라우터 돔 , 리액트 쿼리를 이용하는 모든 컴포넌트를 사용 가능합니다.

만약 `BrowserRouter` 에게 라우터 패스 배열을 넘겨주고 싶다면 스토리를 작성 할 때 다음과 같이 작성해주시면 됩니다.


```tsx
// MyComponent.stories.tsx
import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import MyComponent from './MyComponent';

export default {
  title: 'Example/MyComponent',
  component: MyComponent,
  parameters: {
    memoryRouter: {
      initialEntries: ['/initial/path'],
      initialIndex: 0,
    },
  },
} as ComponentMeta<typeof MyComponent>;

const Template: ComponentStory<typeof MyComponent> = (args) => <MyComponent {...args} />;

export const Default = Template.bind({});
Default.args = {
  // your component props here
};
```




# 첨부 파일 (Optional)
![image](https://github.com/user-attachments/assets/62772e7f-229c-4ded-a916-5469af0a7fc1)

# 레퍼런스 (Optional)

yonghyeun committed:
chore[#52] : withRouter decorator 적용

- 모든 스토리에 react-router-dom 의 기능을 사용 가능하도록 데코레이터 추가
- 문서 : https://storybook.js.org/addons/storybook-addon-react-router-v6

yonghyeun committed:
feat[#52] : login 페이지에 사용할 Icon 추가

yonghyeun committed:
feat[#52] : BackwardNavigationBar 위젯 생성

yonghyeun committed:
feat[#52] : 로그인 HyperLink 추가

yonghyeun committed:
feat[#52] : /login 경로 페이지 생성

yonghyeun committed:
fix[#52] : 발바닥 이모지 제거

yonghyeun committed:
feat[#52] : BackwardNavigationBar onClick props 설정

yonghyeun committed:
feat[#52] : 로그인 페이지 스토리북 추가

yonghyeun committed:
fix[#52] : 로그인 하이퍼링크 entities -> features로 레이어 이동

yonghyeun committed:
chore[#52] : withRouter 패키지 덜어내고 직접 구현하기

yonghyeun committed:
feat[#52] : 세션 유지 정보를 담을 useAuthStore 생성

yonghyeun committed:
feat[#52] : /login 경로 소셜 로그인 버튼 생성

- react query 를 이용하여 로그인 로직 생성
- TODO - 리다이렉션 로직 생성하기

yonghyeun committed:
feat[#52] : /login 경로 로그인 페이지 생성

yonghyeun commented:
### ✅ Test-Storybook Results - Success
```
PASS browser: chromium ../../../../../tmp/e47e26dbc9188171cc3c2a6722129cdd/shared-input.test.js
PASS browser: chromium ../../../../../tmp/e47e26dbc9188171cc3c2a6722129cdd/shared-button.test.js
PASS browser: chromium ../../../../../tmp/e47e26dbc9188171cc3c2a6722129cdd/shared-textarea.test.js

Test Suites: 3 passed, 3 total
Tests:       11 passed, 11 total
Snapshots:   0 total
Time:        4.614 s
Ran all test suites.

```
🚀 [스토리북 확인하기](https://66bae48876a5167719ee1778-jqrdxsprac.chromatic.com)

yonghyeun committed:
chore[#52] : storybook-addon-remix-react-router 패키지 제거

- 해당 패키지를 제거하고 직접 구현했습니다.

yonghyeun commented:
### ✅ Test-Storybook Results - Success
```
PASS browser: chromium ../../../../../tmp/4dd0975a1cf90c68eed3980e0e66473a/shared-button.test.js
PASS browser: chromium ../../../../../tmp/4dd0975a1cf90c68eed3980e0e66473a/shared-textarea.test.js
PASS browser: chromium ../../../../../tmp/4dd0975a1cf90c68eed3980e0e66473a/shared-input.test.js

Test Suites: 3 passed, 3 total
Tests:       11 passed, 11 total
Snapshots:   0 total
Time:        4.152 s
Ran all test suites.

```
🚀 [스토리북 확인하기](https://66bae48876a5167719ee1778-jqrdxsprac.chromatic.com)

aken-you commented on .storybook/preview.tsx:
```diff
@@ -0,0 +1,36 @@
+import React from "react";
+import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
+import { MemoryRouter } from "react-router-dom";
```
MemoryRouter는 이전 페이지의 경로를 기억하기 위해 설정하신 걸까요?

yonghyeun replied:
https://reactrouter.com/en/main/router-components/memory-router

넵 테스트 상황에서 가장 많이 쓰이는 라우터 종류로 `initialEntries` 등에 `window.history` 스택 값처럼 경로들을 지정해줄 수 있더라고요 , 혹은 어떤 경로에서라거나 

그래서 추후 테스트 할 때 특정 경로에서의 테스트나,  라우팅 테스트가 필요한 경우를 대비하여 넣었습니다.

aken-you commented on src/pages/login/login.stories.tsx:
```diff
@@ -0,0 +1,38 @@
+
+const meta: Meta = {
+  title: "Pages/login",
+  tags: ["pages"],
```
`tags: ["pages"]`로 설정하면 어떻게 되나요?

yonghyeun replied:
별 다른 일이 없어서 제거했습니다.

aken-you commented on src/pages/login/login.stories.tsx:
```diff
@@ -0,0 +1,38 @@
+          <Route index element={<LoginPage />} />
+        </Route>
+      </Route>
+    </Routes>
```
혹시 이 페이지 스토리북 작성한 이유가 있을까요?

yonghyeun replied:
스토리북 작성한 이유는 딱히 없습니다. 스토리북을 보고 개발하기 때문에 개발하면서 작성했어요 

다만 만들어두면 나중에 디자이너분이랑 소통 할 떄에도 좋을 거 같네요

aken-you commented on src/features/auth/ui/login-hyperlink.tsx:
```diff
@@ -0,0 +1,75 @@
+
+export const EmailLoginHyperLink = () => (
+  <Link
+    to="/login/email"
```
shared/constants에 있는 값 사용하면 될 거 같아요 👀

aken-you commented on src/features/auth/api/login.ts:
```diff
@@ -0,0 +1,46 @@
+      }
+      const data = await response.json();
+      return data;
+    },
```
이 부분도 나중에 함수로 빼실건가요?!

yonghyeun replied:
넵 이 부분도 `axios` 를 도입하든  공용 함수를 만들든 나중에 변경해야 할 거 같습니다 너무 거추장스러워보이는 거 같아요 ㅎㅎ

aken-you commented on src/features/auth/ui/index.ts:
```diff
@@ -0,0 +1 @@
+export * from "./login-hyperlink";
```
혹시 파일명이 케밥표기법으로 하신 이유가 있나요?

yonghyeun replied:
컴포넌트들은 아니면서 , 로그인 하이퍼링크들을 모아둔 파일이라서 .. 

제가 작명센스가 없습니다. 좋은 이름을 생각해볼게요

aken-you commented on src/features/auth/ui/login-hyperlink.stories.tsx:
```diff
@@ -0,0 +1,20 @@
+
+export const Default: Story = {
+  render: () => <HyperLinkWidget />,
+};
```
이 widget이 로그인 페이지 스토리와 크게 다른 점을 못찾겠네요 🤔 
혹시 작성한 이유가 있을까요?

yonghyeun replied:
이 또한 개발 하면서 같이 작성했습니닷 

우선은 스토리북 만들어두고 나중에 덜어내는건 어떨까요 

디자이너분과 어떻게 소통하게 될지 모르니까요 

개발이 종료될 때 쯤 되면 덜어낼게요

aken-you commented on src/features/auth/ui/login-hyperlink.tsx:
```diff
@@ -0,0 +1,75 @@
+
+export const OAuthLoginHyperLinks = () => {
+  const [OAuthServerName, setOAuthServerName] =
+    useState<OAuthServerName | null>(null);
```
상태 네이밍이 앞에 대문자로 시작하는 이유가 있을까요?

yonghyeun replied:
의식의 흐름대로 .. OAuth 라는 단어가 익숙해서 저렇게 썼나 봅니다

이 부분은 oAuth 로 변경하도록 하겠습니닷

aken-you replied:
OAuth가 더 익숙하네요 bb

yonghyeun committed:
fix[#52] : 리뷰 사항 적용하여 변경

- 카멜케이스 유지하기 OAuth .. -> oAuth
- login-hyperlinks -> hyperlinks 파일명 변경
- 스토리북 tags 에서 pages 제거

yonghyeun commented:
### ✅ Test-Storybook Results - Success
```
PASS browser: chromium ../../../../../tmp/2eaf324b8432692cbd62cd6f789f087a/shared-textarea.test.js
PASS browser: chromium ../../../../../tmp/2eaf324b8432692cbd62cd6f789f087a/shared-input.test.js
PASS browser: chromium ../../../../../tmp/2eaf324b8432692cbd62cd6f789f087a/shared-button.test.js

Test Suites: 3 passed, 3 total
Tests:       11 passed, 11 total
Snapshots:   0 total
Time:        5.458 s
Ran all test suites.

```
🚀 [스토리북 확인하기](https://66bae48876a5167719ee1778-jqrdxsprac.chromatic.com)

yonghyeun committed:
fix[#52] : OAuth 이름 롤백 oauth -> OAuth

yonghyeun commented:
### ✅ Test-Storybook Results - Success
```
PASS browser: chromium ../../../../../tmp/ae6bd41d370f3e5342f095826653f9d1/shared-input.test.js
PASS browser: chromium ../../../../../tmp/ae6bd41d370f3e5342f095826653f9d1/shared-textarea.test.js
PASS browser: chromium ../../../../../tmp/ae6bd41d370f3e5342f095826653f9d1/shared-button.test.js

Test Suites: 3 passed, 3 total
Tests:       11 passed, 11 total
Snapshots:   0 total
Time:        5.353 s
Ran all test suites.

```
🚀 [스토리북 확인하기](https://66bae48876a5167719ee1778-jqrdxsprac.chromatic.com)
